src/Std/Sat/AIG/CNF.lean

```diff
@@ -0,0 +1,721 @@
+  go (aig : AIG Nat) (upper : Nat) (h : upper < aig.decls.size) (state : toCNF.State aig) :
+      { out : toCNF.State aig // toCNF.State.IsExtensionBy state out upper h } :=
+    if hmarked:state.cache.marks[upper]'(by have := state.cache.hmarks; omega) then
+      ⟨state, by apply toCNF.State.IsExtensionBy_rfl <;> assumption⟩
+    else
+      let decl := aig.decls[upper]
+      match heq:decl with
+      | .const b => state.addConst upper h heq
+      | .atom a => state.addAtom upper h heq
+      | .gate lhs rhs linv rinv =>
+        have := aig.inv h heq
+        let ⟨lstate, hlstate⟩ := go aig lhs (by omega) state
+        let ⟨rstate, hrstate⟩ := go aig rhs (by omega) lstate
+
+        have : toCNF.State.IsExtensionBy state rstate lhs (by omega) := by
+          apply toCNF.State.IsExtensionBy_trans_left
+          . exact hlstate
+          . exact hrstate
+
+        let ⟨ret, hretstate⟩ := rstate.addGate upper h heq this.trueAt hrstate.trueAt
+        ⟨
+          ret,
+          by
+            apply toCNF.State.IsExtensionBy_trans_right
+            . exact hlstate
+            . apply toCNF.State.IsExtensionBy_trans_right
+              . exact hrstate
+              . exact hretstate
+        ⟩
+
+/--
+The function we use to convert from CNF with explicit auxiliary variables to just `Nat` variables
+in `toCNF` is an injection.
+-/
+theorem toCNF.inj_is_injection {aig : AIG Nat} (a b : CNFVar aig) :
+    toCNF.inj a = toCNF.inj b → a = b := by
+  intro h
+  cases a with
+  | inl =>
+    cases b with
+    | inl =>
+      dsimp only [inj] at h
+      congr
+      omega
+    | inr rhs =>
+      exfalso
+      dsimp only [inj] at h
+      have := rhs.isLt
+      omega
+  | inr lhs =>
+    cases b with
+    | inl =>
+      dsimp only [inj] at h
+      omega
+    | inr =>
+      dsimp only [inj] at h
+      congr
+      omega
+
+/--
+The node that we started CNF conversion at will always be marked as visited in the CNF cache.
+-/
+theorem toCNF.go_marked :
+    (go aig start h state).val.cache.marks[start]'(by have := (go aig start h state).val.cache.hmarks; omega) = true :=
+  (go aig start h state).property.trueAt
+
+/--
+The CNF returned by `go` will always be SAT at `cnfSatAssignment`.
+-/
+theorem toCNF.go_sat (aig : AIG Nat) (start : Nat) (h1 : start < aig.decls.size) (assign1 : Nat → Bool)
+    (state : toCNF.State aig) :
+    (go aig start h1 state).val.Sat (cnfSatAssignment aig assign1)  := by
+  have := (go aig start h1 state).val.inv assign1
+  rw [State.sat_iff]
+  simp [this]
+
+/--
+Connect SAT results about the CNF to SAT results about the AIG.
+-/
+theorem toCNF.go_as_denote (aig : AIG Nat) (start) (h1) (assign1) :
+    ((⟦aig, ⟨start, h1⟩, assign1⟧ && (go aig start h1 (.empty aig)).val.eval (cnfSatAssignment aig assign1)) = sat?)
+      →
+    (⟦aig, ⟨start, h1⟩, assign1⟧ = sat?) := by
+  intro h
+  have := go_sat aig start h1 assign1 (.empty aig)
+  simp only [State.Sat, CNF.sat_def] at this
+  simpa [this] using h
```

Comment:
The logic here is a bit weird? The first fact here is the interesting one, then it's just logic puzzle?
```
theorem toCNF.go_as_denote' (aig : AIG Nat) (start) (h1) (assign1) :
    ⟦aig, ⟨start, h1⟩, assign1⟧ → (go aig start h1 (.empty aig)).val.eval (cnfSatAssignment aig assign1) := by
  have := go_sat aig start h1 assign1 (.empty aig)
  simp only [State.Sat, CNF.sat_def] at this
  simp [this] 

/--
Connect SAT results about the CNF to SAT results about the AIG.
-/
theorem toCNF.go_as_denote (aig : AIG Nat) (start) (h1) (assign1) :
    ((⟦aig, ⟨start, h1⟩, assign1⟧ && (go aig start h1 (.empty aig)).val.eval (cnfSatAssignment aig assign1)) = sat?)
      →
    (⟦aig, ⟨start, h1⟩, assign1⟧ = sat?) := by
  have := go_as_denote' aig start h1 assign1
  by_cases CNF.eval (cnfSatAssignment aig assign1) (go aig start h1 (State.empty aig)).val.cnf <;> simp_all
```